src/Analysers/LikelihoodRecoAnalyser.cpp

```diff
@@ -0,0 +1,180 @@
+	histMan_->setCurrentHistogramFolder(histogramFolder_);
+	int ttbarpartonsfound = 0;
+	const JetCollection AllJets(event->Jets());
+	int TypeofSolution = 0;//1 = correct, 2=wrong cobination, 3 = not reconstructible, 4 = not semileptonic
+	int EventCorrect = 0;//0=False, 1=True
+
+	LikelihoodBasedTopPairReconstruction likelihoodReco( selectedLepton, met, jets, bjets );
+
+	likelihoodReco.reconstruct();
+
+	TtbarHypothesisCollection allSolutions = likelihoodReco.getAllSolutions();
+
+
+	if (allSolutions.size() != 0){
+
+		for (unsigned int index = 0; index<allSolutions.size(); ++index){
+			TtbarHypothesisPointer solution = allSolutions[index];
+			double Discrim = solution->discriminator;
+			treeMan_->Fill( "likelihoodRatioDiscrimAllJetCombinations", Discrim );
+		}
+
+		TtbarHypothesisPointer bestSolution = likelihoodReco.getBestSolution();
+
+		ParticlePointer TopHad = bestSolution->hadronicTop;
+		ParticlePointer TopLep = bestSolution->leptonicTop;
+		ParticlePointer W = bestSolution->hadronicW;
+		// double Nu = bestSolution->neutrinoChi2;
+
+		treeMan_->Fill("LeptonicTop_Pt",TopLep->pt() );
+		treeMan_->Fill("LeptonicTop_Eta",TopLep->eta() );
+		treeMan_->Fill("LeptonicTop_Energy",TopLep->energy() );
+		treeMan_->Fill("LeptonicTop_Mass",TopLep->mass() );
+
+		treeMan_->Fill("HadronicTop_Pt",TopHad->pt());
+		treeMan_->Fill("HadronicTop_Eta",TopHad->eta() );
+		treeMan_->Fill("HadronicTop_Energy",TopHad->energy());
+		treeMan_->Fill("HadronicTop_Mass",TopHad->mass() );
+
+		// Require one of the event selections to have been satisfied, and we have a genuine ttbar event
+		int selectionCriteria = -1;
+		if ( event->PassesElectronTriggerAndSelection() && event->isSemiLeptonicElectron() ) selectionCriteria = SelectionCriteria::ElectronPlusJetsReference;
+		else if ( event->PassesMuonTriggerAndSelection() && event->isSemiLeptonicMuon() ) selectionCriteria = SelectionCriteria::MuonPlusJetsReference;
+		if ( selectionCriteria < 0 ) {
+			TypeofSolution = 4;
+			treeMan_->Fill("TypeofSolution", TypeofSolution );
+		}
+
+
+		for ( unsigned int AllJetIndex=0; AllJetIndex < AllJets.size(); ++AllJetIndex ) {
+
+			JetPointer AllJet = AllJets[AllJetIndex];
+
+			if ( AllJet->matched_parton() == 0 ) continue;//if null pointer
+
+			if ( AllJet->ttbar_decay_parton() == 3 ){
+				// cout << "Found qbar decay parton" << endl;
+				++ttbarpartonsfound;
+			}
+			if ( AllJet->ttbar_decay_parton() == 4 ){
+				// cout << "Found q decay parton" << endl;
+				++ttbarpartonsfound;
+			}		
+			if ( AllJet->ttbar_decay_parton() == 5 ){
+				// cout << "Found leptonicb decay parton" << endl;
+				++ttbarpartonsfound;
+			}		
+			if ( AllJet->ttbar_decay_parton() == 6 ){
+				// cout << "Found hadronicb decay parton" << endl;
+				++ttbarpartonsfound;
+			}
+		}
+		// cout << "-------------------" << endl;
+		if ( ttbarpartonsfound != 4 ){
+			TypeofSolution = 3;
+			treeMan_->Fill("TypeofSolution", TypeofSolution);
+		}
```

Comment:
It looks like you might be filling the TypeOfSolution branch a number of times per event?  e.g. if the event passes offline selection but isn't semileptonic (so type of solution = 4), you probably won't find all the ttbar_decay_partons, so type of solution is also = 3.  I'm not sure what happens if you call Fill on the same branch twice in one event...

I think what you need to do is :

if ( ttbarpartonsfound != 4 && selectionCriteria > 0 ) ?